### ISSUE TITLE
BUG: transferred NFT did not transferred to recipient, but get escrowed

### ISSUE THREAD

**Issue by taitruong** [game-of-nfts/gon-evidence#517]
Found 2 examples:
1. transfer from Stargaze to Juno
2. transfer back from Stargaze to Irisnet

In both cases ack failed. As a result NFT is not returned/unlocked on source chain (Stargaze):

1. Stargaze -> Juno
a) transfer: https://gon.ping.pub/stargaze/tx/36E5AC4BF6EF7C7522031F9B27DA910BD2C44C14AC082E514596BDBD963E1379
b) ack: https://gon.ping.pub/juno/tx/499CA03C3E91CB958E76431468BC3772513AF12CFE9273C5268A85DEA2338822
error:
- raw log: {"key":"packet_ack","value":"{\"error\":\"codespace: wasm, code: 5\"}
- message: code 5, rpc error: code = NotFound desc = packet acknowledgement hash not found: key not found

2. Stargaze -> IRISnet (back transfer)
a) transfer: https://gon.ping.pub/stargaze/tx/0364AD07BBCF638122F0F35D376F9B27BA91A5E01B37C436CA015D20B0EEC919
transferred succesfully

b) ack: https://gon.ping.pub/iris/tx/F3964D610E30F3CD74353A4BE9191CBEBECCF8D2378D9BED42A89FCE80904408
ack error:
- raw log: {"key":"packet_ack","value":"{\"error\":\"ABCI code: 1: error handling packet: see events for details\"}
- decoding bech32 failed: invalid separator index 11 | decoding bech32 failed: invalid separator index 11

**Comment by taitruong** [comment]
More details on case 1:
i) token was successfully transfer before, starting from IRISnet -> Stargaze -> Juno -> Uptick
ii) succesfully send back from Uptick -> Juno -> Stargaze -> Irisnet
iii) resend Irisnet -> Stargaze -> Juno
Last transfer is where it got stucked on Juno. There is another bug #501 outlining when token got received second time. There it is not clear, what '2nd time' means. 2nd transfer or second packet  receive?

**Comment by taitruong** [comment]
Also bug #497 has another example. maybe this is all related with each other.

**Comment by taitruong** [comment]
> Found 2 examples:
> 
> 1. transfer from Stargaze to Juno
> 2. transfer back from Stargaze to Irisnet
> 
> In both cases ack failed. As a result NFT is not returned/unlocked on source chain (Stargaze):
> 
> 1. Stargaze -> Juno
>    a) transfer: https://gon.ping.pub/stargaze/tx/36E5AC4BF6EF7C7522031F9B27DA910BD2C44C14AC082E514596BDBD963E1379
>    b) ack: https://gon.ping.pub/juno/tx/499CA03C3E91CB958E76431468BC3772513AF12CFE9273C5268A85DEA2338822
>    error:
> 
> * raw log: {"key":"packet_ack","value":"{"error":"codespace: wasm, code: 5"}
> * message: code 5, rpc error: code = NotFound desc = packet acknowledgement hash not found: key not found
> 
> 2. Stargaze -> IRISnet (back transfer)
>    a) transfer: https://gon.ping.pub/stargaze/tx/0364AD07BBCF638122F0F35D376F9B27BA91A5E01B37C436CA015D20B0EEC919
>    transferred succesfully
> 
> b) ack: https://gon.ping.pub/iris/tx/F3964D610E30F3CD74353A4BE9191CBEBECCF8D2378D9BED42A89FCE80904408 ack error:
> 
> * raw log: {"key":"packet_ack","value":"{"error":"ABCI code: 1: error handling packet: see events for details"}
> * decoding bech32 failed: invalid separator index 11 | decoding bech32 failed: invalid separator index 11

On case 2: as noted by sender here: https://discord.com/channels/669268347736686612/1074987031270268958/1085843405940195369 (Cosmos Network, gon-testnet channel)
Token was before alsos send from Irisnet -> SG, then it failed on sending back from SG -> Irisnet.

**Comment by gjermundgaraba** [comment]
I am pretty sure #497 is related here

**Comment by taitruong** [comment]
![image](https://user-images.githubusercontent.com/776986/225602780-4c3350bb-7e8c-4c7b-818c-048f3c474bd5.png)

**Comment by taitruong** [comment]
For reproducing the bug, I simply created a new collection. So bug occurs for any NFT and collection.

**Comment by beal2912** [comment]
Hello, 

I have replicated the problem with my second airdrop : gir2/beal2912  on Flow F31

What I did was 

OK : i - 2 - s   + gon : 57D4D11C47CC5FF626459C51D38065C675E48D4882E9733BC0907A7C00238B84
OK : s - 2 - J + gon : 2D06F9E97C0142A8FEE52C02B639608A37F70944061FFED27845E58D25B50AA9

then roll back :
OK : j - 2 - s + gon : 8DAC483CED9C5D34BA1F535E2431C89493BBE9DA6BC38CAD21D571537F025600
OK : s - 2 - i + gon : 6B6CB5E4C4140F420B65263F275F18EED051F34C9C57EF09253954D100BD184F

then restart :

KO : i - 2 - s + gon : 161C06F168CD1FFEC3C7414DF7DB28F16F01D6BBE6586ACF8A0A0B7D0F542948
the tx is successful but with "packet acknowledgement hash not found" 

my assumption is that I used the path a 2 times with self relay with GON,  and the 2nd time there is a mix up between acknowledgement packet but the tx is successul so the nft stay in the escrow address,

**Comment by taitruong** [comment]
I also tried with other chains: J(uno), U(ptick), O(mniflix), and I(risnet)

1. visits
J -> U -> O -> I
2. back visit:
I -> O -> U -> J
3. Revisit
J -> U -> O -> I

But couldn' reproduce it here. Only when Stargaze and Juno is used. This might be related to this spam attack on channel between Stargaze and Juno: #479

**Comment by taitruong** [comment]
I'm wondering whether it has something to do with packet sequences. I dunno how numbers are generated. But once I had a packet sequence 14005 (transfer from Juno to Stargaze) and later a lower sequence number of 282. As noted in #479 malicious attack transferred ~500k NFTs.

**Comment by giansalex** [comment]
> then restart :
KO : i - 2 - s + gon : 161C06F168CD1FFEC3C7414DF7DB28F16F01D6BBE6586ACF8A0A0B7D0F542948
the tx is successful but with "packet acknowledgement hash not found"

Your last tx got timeout: [5D9FD20013EF5A97CD1960F6C6A1F8996A4C49FC6B95260F50D2A12E53CEDE76](https://gon.ping.pub/iris/tx/5D9FD20013EF5A97CD1960F6C6A1F8996A4C49FC6B95260F50D2A12E53CEDE76)

msg: `packet acknowledgement hash not found`, it's not part of the transaction result, it's from the explorer, is causing confusion

**Comment by taitruong** [comment]
@giansalex why do you think it is a timeout, on dest chain checking this: https://gon.ping.pub/stargaze/tx/26D361FCE11E66FF2A30DE08E1EE70899549C1A054D936D8FF7B62C2C1C10500

tx height : 3885887
timestamp: 2023-03-16T12:08:24Z
timeout height is: 3886875 (later than height)
timeout timespam: 1678969095276252876 (10 min later than timestamp)

**Comment by giansalex** [comment]
> @giansalex why do you think it is a timeout? On dest chain checking this: https://gon.ping.pub/stargaze/tx/26D361FCE11E66FF2A30DE08E1EE70899549C1A054D936D8FF7B62C2C1C10500
> 
> tx height : 3885887 timestamp: 2023-03-16T12:08:24Z timeout height is: 3886875 (later than height) timeout timespam: 1678969095276252876 (10 min later than timestamp)

That was the result of using gon-cli, maybe a issue there.

**Comment by giansalex** [comment]
@taitruong I tried again and Timeout Tx is accepted, even though ACK was already received. Although this does not affect the token owner, it is probably related to the lack of [IBC Ante](https://github.com/cosmos/ibc-go/blob/v5.0.1/modules/core/ante/ante.go).

Transfer: 161C06F168CD1FFEC3C7414DF7DB28F16F01D6BBE6586ACF8A0A0B7D0F542948
RCV: 26D361FCE11E66FF2A30DE08E1EE70899549C1A054D936D8FF7B62C2C1C10500
ACK: 88CC2FFAC28489A4BA5E7134392C249322DC795E6DC5FF778B566611AAE43084

Timeout:
5D9FD20013EF5A97CD1960F6C6A1F8996A4C49FC6B95260F50D2A12E53CEDE76
32D7207D342190FBE9B758AD3C93B7769659F56D13B8A26BF5850C3842AEDC88

**Comment by taitruong** [comment]
# Summary
Collection originally minted on IRISnet. Transfer through various chains: I(RISnet), S(targaze), J(uno), U(ptick), O(mniflix)

Alright summing things up. Problem is described here: #501:

tl;dr: retransferring (backtrack of an) NFT in ICS721 wasm contract, burns on ACK response ("locked" chain until then). As long as  ACK is not send/relayed yet, it keeps locked. Though on other ("active) chain it is active/unlocked. Sending an NFT from "active" to "locked" chain is not possible, since NFT is still there and not yet burned. Once ACK has been send, NFT can be send to "burned" chain.

# Problem
token gets transferred throught these chains:

1. A -> B: NFT minted on chain B and locked on chain A
2. B -> A: nft is unlocked on chain a, token is not getting burned on chain b, because ack  does not arrive
3: A -> B: token is locked on chain a and transferred to chain b, but can't because it is already there.

For 3) NFT results can be:
a) active on chain A, locked-not-yet-burned on chain B
b) locked-not-yet-unlocked on both chain A and locked-not-yet-burned on chain B


Examples below.

# Active on chain A, Locked-Not-Yet-Burned on chain B

Affected chains: IRISnet and Stargaze

For **a) active on chain A, locked-not-yet-burned on chain B**
- token was sent I > S and then back  to > I
- result: Now token is active on I, but locked on S

user tried sending from I back to S, but it failed ("error":"codespace: wasm, code: 5") and got reverted/unlocked back to I
multiple attempts afterwards I > S, it all failed:

transfer from I: https://gon.ping.pub/iris/tx/92E4C554A216237B0880DFA3D47A995198A71F279A375FA5FABB2000F7E401BC
ack error from S: https://gon.ping.pub/stargaze/tx/D67326D58BAD39CC169A47989DB4A44294F5B723349A32527EFE304D29BA4E37

transfer from I: https://gon.ping.pub/iris/tx/88BC117968E47E2B82C00D5B76317F3035A3A517DB7F9C00324CB36B6877C0D0
ack error from S: https://gon.ping.pub/stargaze/tx/803DD8471D12B2E9D32CBA0972CB9C43346DC06F7733A2FB5929E6736D342004

transfer from I: https://gon.ping.pub/iris/tx/BEFCFEFAADCECEFE8B48A048F1B3DDF2613EBDC6A43C1D50F1984D4C1D77FF14
ack error from S: https://gon.ping.pub/stargaze/tx/2A930AE1C93857D9B9D1A71718CB3C9C3270316D194961F91965BDA141ACEF58

transfer from I: https://gon.ping.pub/iris/tx/5EC4873825477B6773EEB1061FD57B7AB36EA4E496295E371A5C3D6AC7419263
ack error from S: https://gon.ping.pub/stargaze/tx/ABA9831C16CFEE0BD96B6EF63CF0ECE17F13EF7C46D923A3E682F9E5F38BCD3C

# Locked-Not-Yet-Unlocked on both chain A and Locked-Not-Yet-Burned on chain B

Affected chains: Stargaze and Juno

This is prob the worst case:
1. NFT got send I > S > J
2. send back J > S > I
3. resend I > S > J

When it got send back in 2) from J > S, **NFT got locked-not-yet-burned on Juno.**
In 3) when sending from S > J, it failed - since it is still locked and not burned on Juno, in addition ack fail from J > S got stucked, so on S side **NFT is locked-not-yet-unlocked** since ack hasn't arrived yet.

transfer from Stargaze:
https://gon.ping.pub/stargaze/tx/36E5AC4BF6EF7C7522031F9B27DA910BD2C44C14AC082E514596BDBD963E1379
ack error from Juno: https://gon.ping.pub/juno/tx/499CA03C3E91CB958E76431468BC3772513AF12CFE9273C5268A85DEA2338822

**Comment by taitruong** [comment]
Due to the Spam between Stargaze and Juno on a specific channel, there are right now ~470k open packets. Even though NFTs are send now through different channels relayers got stucked and are not able to relay all packets.

We know that the root cause is that ACK is not send  yet.

@gjermundgaraba, it is possible that gon cli's self relaying feature is "hiding" this problem, because self-relay:
 picks a specific packet sequence to relay. at least it relays recv package, but 99% sure ack is NOT relayed.
 
 Why? Once NFT passed **either** Stargaze **or** Juno, ACKs gets stucked! 
 
 Normally relayers puts new packets in queue. Normal use case is that for congested chains like Stargaze and Juno, all incoming packets are queued. So transferred NFT gets queued, but not transferred (minted or burned) on Stargaze or Juno. User would than immediately see that NFT isn't there yet. With self-relay NFT at least can pass ONE time, next time sending back and self-relay again, it fails immediately because NFT is not yet burned.

**Comment by giansalex** [comment]
The problem is that token burning should occur at Transfer time, not at the IBC ACK. 

will be reviewed next week https://github.com/game-of-nfts/gon-evidence/issues/501#issuecomment-1472522257

**Comment by taitruong** [comment]
> The problem is that token burning should occur at Transfer time, not at the IBC ACK.
> 
> > will be reviewed next week [#501 (comment)](https://github.com/game-of-nfts/gon-evidence/issues/501#issuecomment-1472522257)

Not sure whether this solves this, because of this case:
back transfer from A to B, e.g with wrong/incorrect recipient address.

normally on chain B it sends an ACK fail for rollback, but since ACK gets stucked, NFT is neither on chain B nor on chain A. Once ACK fail has arrived on chain B, it got reminted and returned back to sender.

**Comment by giansalex** [comment]
> Not sure whether this solves this, because of this case: back transfer from A to B, e.g with wrong/incorrect recipient address.
> 
> normally on chain B it sends an ACK fail for rollback, but since ACK gets stucked, NFT is neither on chain B nor on chain A. Once ACK fail has arrived on chain B, it got reminted and returned back to sender.

if the NFT is neither on chain A nor B, the user must wait for the relayers. This is a normal case that also occurs when the packet is in Timeout and the relayers are not operational.

**Comment by ElfenLied2019** [comment]
I have tried multiple times from i -- (1) -->s -- (1) -->o -- (1) -->j -- (1) -->i, and each time I lose nft at o -- (1) -->j

https://gon.ping.pub/omniflix/tx/845A21636100E5644A7C50F00D5AEB64D3BD5ACC1E4C87E83785AAC445A79DDC
https://gon.ping.pub/omniflix/tx/EE3F6EE8E874B633C0985C32659953942E137C94A543B6914DA870F177E7F2AB

**Comment by taitruong** [comment]
> > Not sure whether this solves this, because of this case: back transfer from A to B, e.g with wrong/incorrect recipient address.
> > normally on chain B it sends an ACK fail for rollback, but since ACK gets stucked, NFT is neither on chain B nor on chain A. Once ACK fail has arrived on chain B, it got reminted and returned back to sender.
> 
> if the NFT is neither on chain A nor B, the user must wait for the relayers. This is a normal case that also occurs when the packet is in Timeout and the relayers are not operational.
> 
> in case #501, the remote wasm chain cannot accept the NFT, due to an internal error, unexpected behavior

There is no internal error. Receiving chain either sends an ack success or ack fail. Acks are always relayed. Read here: https://medium.com/the-interchain-foundation/everything-you-need-to-know-about-ibc-channels-faqs-part-2-44249a7b7e0b

```
Do acknowledgements have timeouts?

No, they do not. There needs to be at least one message that can be submitted async. Otherwise it causes a circular problem where if timeouts existed for acknowledgements, then the message proving that the acknowledgment timed out could itself timeout, and so on — causing an endless loop.
```

**Comment by giansalex** [comment]
> There is no internal error. Receiving chain either sends an ack success or ack fail. Acks are always relayed. Read here: https://medium.com/the-interchain-foundation/everything-you-need-to-know-about-ibc-channels-faqs-part-2-44249a7b7e0b

I am referring to the error that happens when cw-ics721 executes the [cw721 contract](https://gon.ping.pub/stargaze/tx/2A930AE1C93857D9B9D1A71718CB3C9C3270316D194961F91965BDA141ACEF58) on IBC Recv event.
```
{"error":"codespace: wasm, code: 5"}
```

This error should not occur if cw-ics721 would faithfully follow the [ICS721 spec](https://github.com/cosmos/ibc/tree/main/spec/app/ics-721-nft-transfer).

**Comment by taitruong** [comment]
> > There is no internal error. Receiving chain either sends an ack success or ack fail. Acks are always relayed. Read here: https://medium.com/the-interchain-foundation/everything-you-need-to-know-about-ibc-channels-faqs-part-2-44249a7b7e0b
> 
> I am referring to the error that happens when cw-ics721 executes the _cw721 contract_ on [IBC Recv event](https://gon.ping.pub/stargaze/tx/2A930AE1C93857D9B9D1A71718CB3C9C3270316D194961F91965BDA141ACEF58).
> 
> ```
> {"error":"codespace: wasm, code: 5"}
> ```
> 
> This error should not occur if cw-ics721 would faithfully follow the [ICS721 spec](https://github.com/cosmos/ibc/tree/main/spec/app/ics-721-nft-transfer).

There has been a long discussion about this between IRISnet, Stargaze and Juno. The spec shows one possible solution. In this case to burn on transfer while sending back. Another solution is burn on ACK success. Technically both solutions are fine.

From functional/business perspective, the question is: Do you really want to burn your NFT on chain B, while complete flow is not finished yet? As long as chain A is not finished. In the meanwhile for the holder its intermediate state is like this:

- holder does not hold it anymore, since it is burned on chain B
- NFT is escrowed on chain A

While in the other case NFT is escrowed on both chains. It might be subtle, but during GoN a lot suffered from this. And most of participants had the feeling their NFT got lost and they couldn't find it on chain B anymore. They saw nothing on Chain has happend and it is still escrowed.

It is just my opinion, but technically both fulfills the same purpose and leads to the same result. But from user's perspective it is not only a technical issue, they get the impression of having their NFT lost.

Users start panicking, don't understand what is going on, thinking about they got ruggeg or whatever. For them it is just gone.

**Comment by taitruong** [comment]
After all, at least we have identified the source of this: packets have not been send, first implementation of self-relay only relayed transfer packet, but not ack packet. Now with self-relay at least all packets for given transfer has been relayed.

So without relay, it lead NFT being:
- burned on chain B
- escrowed on chain A

With relaying only transfer packet, it led to same result, but it made it worst when doing a re-transfer, so complete flow is like:
1. I transfer to J
2. J back transfer to I (ack not relayed)
3. I re-transfer to J (2nd ack not  relayed

When doing 3. it ended up NFT being escrowed on both sides. Instead of burning during transfer, but escrow it first and then burns in on ACK succuss. User could have gone only through step 2 - without self-relaying. This is the root of cause. In other case when getting stuck doing step 3, the root of cause is not obvious and much harder to find!
